### PR TITLE
*: add fail over scheduler

### PR DESCRIPTION
**Pull request by nolouch**
now checker couldn't found abnormal region immediately. and there are some priorities for scheduling the fail store regions.

### CODE REVIEW

File: server/core/region.go [hunk 1]
```diff
@@ -581,13 +605,17 @@ func (r *RegionsInfo) GetRegionStats(startKey, endKey []byte) *RegionStats {
 
 const randomRegionMaxRetry = 10
 
-func randRegion(regions *regionMap) *RegionInfo {
+func randRegion(regions *regionMap, skipUnhealth bool) *RegionInfo {
```
Comment (disksing) [inline server/core/region.go]: I think we can pass a select/filter function to select the region(s) we want

**Comment by Connor1996** [comment]
There is a new way to select region(s). Please merge master and update your code.

File: server/config.go [hunk 2]
```diff
@@ -398,6 +398,7 @@ var defaultSchedulers = SchedulerConfigs{
 	{Type: "balance-region"},
 	{Type: "balance-leader"},
 	{Type: "hot-region"},
+	{Type: "fail-over"},
```
Comment (Connor1996) [inline server/config.go]: after upgrading from existed pd, the scheduler won't be added automatically. It is important to notice ops.

File: server/schedulers/fail_over.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+func (f *failOverScheduler) handleDownStore(cluster schedule.Cluster, store *core.StoreInfo) *schedule.Operator {
+	region := cluster.RandLeaderRegion(store.GetId())
+	if region != nil {
+		log.Warnf("[region %d] Down Store [%v] have leader region, Maybe have network problem.", region.GetId(), store)
```
Comment (Connor1996) [inline server/schedulers/fail_over.go]: have leader region -> have no leader region

File: server/schedulers/fail_over.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+		return nil
+	}
+
+	peer := region.GetStorePeer(store.GetId())
+	if peer == nil {
+		return nil
+	}
+	downPeer := region.GetDownPeer(peer.GetId())
+	if downPeer == nil {
+		log.Warnf("[region %d] peer %v not down in down store", region.GetId(), peer)
```
Comment (Connor1996) [inline server/schedulers/fail_over.go]: show the down store id

File: server/schedulers/fail_over.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+		log.Errorf("[region %d] region unhealth: %s", region.GetId(), region)
+		schedulerCounter.WithLabelValues(f.GetName(), "unhealth").Inc()
+		return nil
+	}
+
+	peer := region.GetStorePeer(store.GetId())
+	if peer == nil {
+		return nil
+	}
+	downPeer := region.GetDownPeer(peer.GetId())
```
Comment (Connor1996) [inline server/schedulers/fail_over.go]: is it possible that peer down but store is not down?

Comment (nolouch) [reply server/schedulers/fail_over.go]: No one guarantees this, down peer reported by TiKV, and down store judge by PD.

**Comment by siddontang** [comment]
how about the effect when the store is failed?

**Comment by nolouch** [comment]
close by low priority and inactive.

**Pull request closed without merging**